File: citadel-first/backend/app/schemas/beneficiary.py
from datetime import date, datetime
from decimal import Decimal
from enum import Enum

from pydantic import BaseModel, Field


class BeneficiaryType(str, Enum):
    PRE_DEMISE = "pre_demise"
    POST_DEMISE = "post_demise"


class RelationshipToSettlor(str, Enum):
    SELF = "SELF"
    SPOUSE = "SPOUSE"
    CHILD = "CHILD"
    PARENT = "PARENT"
    SIBLING = "SIBLING"
    PARENTS = "PARENTS"
    GRANDPARENT = "GRANDPARENT"
    GRAND_SON = "GRAND_SON"
    GRAND_DAUGHTER = "GRAND_DAUGHTER"
    NIECE = "NIECE"
    NEPHEW = "NEPHEW"
    PARTNER = "PARTNER"
    FIANCE = "FIANCE"
    FRIEND = "FRIEND"
    GUARDIAN = "GUARDIAN"
    GOD_PARENT = "GOD_PARENT"
    MOTHER_IN_LAW = "MOTHER_IN_LAW"
    FATHER_IN_LAW = "FATHER_IN_LAW"
    SON_IN_LAW = "SON_IN_LAW"
    DAUGHTER_IN_LAW = "DAUGHTER_IN_LAW"
    ASSOCIATE = "ASSOCIATE"
    ADMINISTRATOR = "ADMINISTRATOR"


class BeneficiaryCreateRequest(BaseModel):
    beneficiary_type: BeneficiaryType
    same_as_settlor: bool = False

    # Personal details
    full_name: str | None = None
    nric: str | None = None
    id_number: str | None = None
    gender: str | None = None
    dob: date | None = None
    relationship_to_settlor: str | None = None

    # Address & Contact
    residential_address: str | None = None
    mailing_address: str | None = None
    email: str | None = None
    contact_number: str | None = None

    # Bank details
    bank_account_name: str | None = None
    bank_account_number: str | None = None
    bank_name: str | None = None
    bank_swift_code: str | None = None
    bank_address: str | None = None

    # Share percentage
    share_percentage: Decimal | None = Field(default=None, gt=0, le=100, max_digits=5, decimal_places=2)

    # Attachment S3 keys (set after upload)
    settlor_nric_key: str | None = None
    proof_of_address_key: str | None = None
    beneficiary_id_key: str | None = None
    bank_statement_key: str | None = None


class BeneficiaryUpdateRequest(BaseModel):
    same_as_settlor: bool | None = None
    full_name: str | None = None
    nric: str | None = None
    id_number: str | None = None
    gender: str | None = None
    dob: date | None = None
    relationship_to_settlor: str | None = None
    residential_address: str | None = None
    mailing_address: str | None = None
    email: str | None = None
    contact_number: str | None = None
    bank_account_name: str | None = None
    bank_account_number: str | None = None
    bank_name: str | None = None
    bank_swift_code: str | None = None
    bank_address: str | None = None
    share_percentage: Decimal | None = Field(default=None, gt=0, le=100, max_digits=5, decimal_places=2)
    settlor_nric_key: str | None = None
    proof_of_address_key: str | None = None
    beneficiary_id_key: str | None = None
    bank_statement_key: str | None = None


class BeneficiaryResponse(BaseModel):
    model_config = {"from_attributes": True}

    id: int
    app_user_id: int
    beneficiary_type: str
    same_as_settlor: bool

    full_name: str | None
    nric: str | None
    id_number: str | None
    gender: str | None
    dob: date | None
    relationship_to_settlor: str | None

    residential_address: str | None
    mailing_address: str | None
    email: str | None
    contact_number: str | None

    bank_account_name: str | None
    bank_account_number: str | None
    bank_name: str | None
    bank_swift_code: str | None
    bank_address: str | None

    share_percentage: Decimal | None

    settlor_nric_key: str | None
    proof_of_address_key: str | None
    beneficiary_id_key: str | None
    bank_statement_key: str | None

    created_at: datetime | None
    updated_at: datetime | None


class BeneficiaryListResponse(BaseModel):
    beneficiaries: list[BeneficiaryResponse]
    has_pre_demise: bool
    has_post_demise: bool


# Fix forward reference
from datetime import datetime  # noqa: E402

BeneficiaryResponse.model_rebuild()

File: citadel-first/mobile/lib/models/beneficiary.dart
class Beneficiary {
  final int id;
  final int appUserId;
  final String beneficiaryType; // "pre_demise" or "post_demise"
  final bool sameAsSettlor;
  final String? fullName;
  final String? nric;
  final String? idNumber;
  final String? gender;
  final DateTime? dob;
  final String? relationshipToSettlor;
  final String? residentialAddress;
  final String? mailingAddress;
  final String? email;
  final String? contactNumber;
  final String? bankAccountName;
  final String? bankAccountNumber;
  final String? bankName;
  final String? bankSwiftCode;
  final String? bankAddress;
  final double? sharePercentage;
  final String? settlorNricKey;
  final String? proofOfAddressKey;
  final String? beneficiaryIdKey;
  final String? bankStatementKey;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  const Beneficiary({
    required this.id,
    required this.appUserId,
    required this.beneficiaryType,
    this.sameAsSettlor = false,
    this.fullName,
    this.nric,
    this.idNumber,
    this.gender,
    this.dob,
    this.relationshipToSettlor,
    this.residentialAddress,
    this.mailingAddress,
    this.email,
    this.contactNumber,
    this.bankAccountName,
    this.bankAccountNumber,
    this.bankName,
    this.bankSwiftCode,
    this.bankAddress,
    this.sharePercentage,
    this.settlorNricKey,
    this.proofOfAddressKey,
    this.beneficiaryIdKey,
    this.bankStatementKey,
    this.createdAt,
    this.updatedAt,
  });

  factory Beneficiary.fromJson(Map<String, dynamic> json) {
    return Beneficiary(
      id: json['id'] as int,
      appUserId: json['app_user_id'] as int,
      beneficiaryType: json['beneficiary_type'] as String,
      sameAsSettlor: json['same_as_settlor'] as bool? ?? false,
      fullName: json['full_name'] as String?,
      nric: json['nric'] as String?,
      idNumber: json['id_number'] as String?,
      gender: json['gender'] as String?,
      dob: json['dob'] != null ? DateTime.parse(json['dob'] as String) : null,
      relationshipToSettlor: json['relationship_to_settlor'] as String?,
      residentialAddress: json['residential_address'] as String?,
      mailingAddress: json['mailing_address'] as String?,
      email: json['email'] as String?,
      contactNumber: json['contact_number'] as String?,
      bankAccountName: json['bank_account_name'] as String?,
      bankAccountNumber: json['bank_account_number'] as String?,
      bankName: json['bank_name'] as String?,
      bankSwiftCode: json['bank_swift_code'] as String?,
      bankAddress: json['bank_address'] as String?,
      sharePercentage: _parseDouble(json['share_percentage']),
      settlorNricKey: json['settlor_nric_key'] as String?,
      proofOfAddressKey: json['proof_of_address_key'] as String?,
      beneficiaryIdKey: json['beneficiary_id_key'] as String?,
      bankStatementKey: json['bank_statement_key'] as String?,
      createdAt: json['created_at'] != null ? DateTime.parse(json['created_at'] as String) : null,
      updatedAt: json['updated_at'] != null ? DateTime.parse(json['updated_at'] as String) : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      if (beneficiaryType.isNotEmpty) 'beneficiary_type': beneficiaryType,
      'same_as_settlor': sameAsSettlor,
      if (fullName != null) 'full_name': fullName,
      if (nric != null) 'nric': nric,
      if (idNumber != null) 'id_number': idNumber,
      if (gender != null) 'gender': gender,
      if (dob != null) 'dob': dob!.toIso8601String().split('T')[0],
      if (relationshipToSettlor != null) 'relationship_to_settlor': relationshipToSettlor,
      if (residentialAddress != null) 'residential_address': residentialAddress,
      if (mailingAddress != null) 'mailing_address': mailingAddress,
      if (email != null) 'email': email,
      if (contactNumber != null) 'contact_number': contactNumber,
      if (bankAccountName != null) 'bank_account_name': bankAccountName,
      if (bankAccountNumber != null) 'bank_account_number': bankAccountNumber,
      if (bankName != null) 'bank_name': bankName,
      if (bankSwiftCode != null) 'bank_swift_code': bankSwiftCode,
      if (bankAddress != null) 'bank_address': bankAddress,
      if (sharePercentage != null) 'share_percentage': sharePercentage,
      if (settlorNricKey != null) 'settlor_nric_key': settlorNricKey,
      if (proofOfAddressKey != null) 'proof_of_address_key': proofOfAddressKey,
      if (beneficiaryIdKey != null) 'beneficiary_id_key': beneficiaryIdKey,
      if (bankStatementKey != null) 'bank_statement_key': bankStatementKey,
    };
  }

  bool get isPreDemise => beneficiaryType == 'pre_demise';
  bool get isPostDemise => beneficiaryType == 'post_demise';
}

class BeneficiaryListResult {
  final List<Beneficiary> beneficiaries;
  final bool hasPreDemise;
  final bool hasPostDemise;

  const BeneficiaryListResult({
    required this.beneficiaries,
    required this.hasPreDemise,
    required this.hasPostDemise,
  });

  factory BeneficiaryListResult.fromJson(Map<String, dynamic> json) {
    return BeneficiaryListResult(
      beneficiaries: (json['beneficiaries'] as List)
          .map((b) => Beneficiary.fromJson(b as Map<String, dynamic>))
          .toList(),
      hasPreDemise: json['has_pre_demise'] as bool,
      hasPostDemise: json['has_post_demise'] as bool,
    );
  }
}

double? _parseDouble(dynamic value) {
  if (value == null) return null;
  if (value is num) return value.toDouble();
  if (value is String) return double.tryParse(value);
  return null;
}